src/charm.py

```diff
@@ -4,74 +4,228 @@
+        return True
+
+    def _check_kube_control(self, event):
+        self.unit.status = MaintenanceStatus("Evaluating kubernetes authentication.")
+        evaluation = self.kube_control.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        if not self.kube_control.get_auth_credentials(self.unit.name):
+            self.unit.status = WaitingStatus("Waiting for kube-control: unit credentials")
+            return False
+        self.kube_control.create_kubeconfig(
+            self._ca_cert_path, self._kubeconfig_path, "root", self.unit.name
         )
-        self.unit.status = ActiveStatus()
-        self._set_version()
+        return True
+
+    def _check_certificates(self, event):
+        self.unit.status = MaintenanceStatus("Evaluating certificates.")
+        evaluation = self.certificates.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        self._ca_cert_path.write_text(self.certificates.ca)
+        return True
+
+    def _check_config(self):
+        self.unit.status = MaintenanceStatus("Evaluating charm config.")
+        evaluation = self.charm_config.evaluate()
+        if evaluation:
+            self.unit.status = BlockedStatus(evaluation)
+            return False
+        return True
+
+    def _merge_config(self, event):
+        if not self._check_integrator(event):
+            return
+
+        if not self._check_certificates(event):
+            return
+
+        if not self._check_kube_control(event):
+            return
+
+        if not self._check_config():
+            return
+
+        self.unit.status = MaintenanceStatus("Evaluating Manifests")
+        new_hash = 0
+        for controller in self.collector.manifests.values():
+            evaluation = controller.evaluate()
+            if evaluation:
+                self.unit.status = BlockedStatus(evaluation)
+                return
+            new_hash += controller.hash()
+
+        self.stored.deployed = False
+        if self._install_or_upgrade(event, config_hash=new_hash):
+            self.stored.config_hash = new_hash
+            self.stored.deployed = True
+
+    def _install_or_upgrade(self, event, config_hash=None):
+        if self.stored.config_hash == config_hash:
+            log.info("Skipping until the config is evaluated.")
+            return True
 
-    def _set_version(self, event=None):
-        if self.unit.is_leader():
-            self.unit.set_workload_version(self.version)
+        self.unit.status = MaintenanceStatus("Deploying Cinder Storage")
+        self.unit.set_workload_version("")
```

Comment:
```suggestion
        self.unit.set_workload_version(self.collector.short_version)
```
I think the version is missing. Am I overlooking somethig here?

Reply:
that's done on the `_update_status` hook if the charm is successfully deployed.  The intent here is to clear it during an upgrade/install and set it after a successful install in `update_status`